### PR TITLE
Add --doc option

### PR DESCRIPTION
**Pull request by csh519**
## Description

Add `--doc` option to support document scanning.
It supports private config and rule-sets for document scanning.

- Add `--doc` option
- Add test cases for `--doc` option
- Clean code

## How has this been tested?

- [X] Local tested

### CODE REVIEW

File: credsweeper/app.py [hunk 1]
```diff
@@ -69,38 +70,74 @@ def __init__(self,
...
+        config_dict = self._get_config_dict(config_path, api_validation, use_filters, find_by_ext, depth, doc,
+                                            size_limit, exclude_lines, exclude_values)
+        doc_config_dict = self._get_config_dict(None, api_validation, use_filters, find_by_ext, depth, doc, size_limit,
+                                                exclude_lines, exclude_values, "doc_config.json")
+        self.config = Config(config_dict)
+        self.doc_config = Config(doc_config_dict)
+        self.scanner = Scanner(self.config, rule_path)
+        self.doc_scanner = Scanner(self.config, self._get_doc_rule_path())
+        self.deep_scanner = DeepScanner(self.config, self.scanner)
+        self.deep_doc_scanner = DeepScanner(self.doc_config, self.doc_scanner)
```
Comment (babenek) [inline credsweeper/app.py:89]:
```suggestion
        self.doc_scanner = DeepScanner(self.doc_config, self.doc_scanner)
```

Comment (csh519) [reply credsweeper/app.py:89]: Actually there is `self.doc_scanner` above. :cry:

File: tests/filters/test_value_array_dictionary_check.py [hunk 2]
```diff
@@ -16,7 +16,8 @@ def token_rule(self, config) -> Rule:
             "values": ["token"],
             "filter_type": "",
             "use_ml": True,
-            "validations": []
+            "validations": [],
+            "usage_list": ["src", "doc"]
```
Comment (babenek) [inline tests/filters/test_value_array_dictionary_check.py:20]:
```suggestion
            "usage_list": ["src", "doc"],
            "validations": []
```

File: tests/filters/test_value_similarity_check.py [hunk 3]
```diff
@@ -16,7 +16,8 @@ def password_rule(self, config) -> Rule:
             "values": ["password|passwd|pwd"],
             "filter_type": "",
             "use_ml": True,
-            "validations": []
+            "validations": [],
+            "usage_list": ["src", "doc"]
```
Comment (babenek) [inline tests/filters/test_value_similarity_check.py:20]:
```suggestion
            "usage_list": ["src", "doc"],
            "validations": []
```

File: credsweeper/app.py [hunk 4]
```diff
@@ -247,10 +274,12 @@ def file_scan(self, content_provider: Union[DiffContentProvider, TextContentProv
...
         else:
-            # deep scan with possibly data representation
             if self.config.depth:
-                new_size_limit = self.config.size_limit if self.config.size_limit else RECURSIVE_SCAN_LIMITATION
-                candidates = self.deep_scanner.scan(content_provider, self.config.depth, new_size_limit)
+                # deep scan with possibly data representation
+                candidates = self.deep_scanner.scan(content_provider, self.config.depth, self.config.size_limit)
+            elif self.config.doc:
+                # document-specific scanning
+                candidates = self.deep_doc_scanner.scan(content_provider, 0, self.config.size_limit)
```
Comment (babenek) [inline credsweeper/app.py:282]: doc_scanner?

File: credsweeper/deep_scanner/deep_scanner.py [hunk 5]
```diff
@@ -73,14 +74,16 @@ def get_deep_scanners(data: bytes) -> List[Any]:
 
     # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
 
-    def scan(self, content_provider: ContentProvider, depth: int, recursive_limit_size: int) -> List[Candidate]:
+    def scan(self, content_provider: ContentProvider, depth: int,
+             recursive_limit_size: Optional[int]) -> List[Candidate]:
```
[the remark this answers is not in the document]
Comment (csh519) [reply credsweeper/deep_scanner/deep_scanner.py:78]: Thank you for review.

I fixed it with ee9b6ec.
Please check.

Comment (babenek) [inline credsweeper/deep_scanner/deep_scanner.py:78]:
```suggestion
             recursive_limit_size: Optional[int] = None) -> List[Candidate]:
```

File: credsweeper/deep_scanner/deep_scanner.py [hunk 6]
```diff
@@ -73,14 +74,16 @@ def get_deep_scanners(data: bytes) -> List[Any]:
...
         """Initial scan method to launch recursive scan. Skips ByteScanner to prevent extra scan
 
             Args:
                 content_provider: ContentProvider that might contain raw data
                 depth: maximal level of recursion
                 recursive_limit_size: maximal bytes of opened files to prevent recursive zip-bomb attack
         """
+        recursive_limit_size = recursive_limit_size if recursive_limit_size is None else RECURSIVE_SCAN_LIMITATION
```
Comment (babenek) [inline credsweeper/deep_scanner/deep_scanner.py:86]:
```suggestion
        recursive_limit_size = recursive_limit_size if isinstance(recursive_limit_size, int) else RECURSIVE_SCAN_LIMITATION
```

File: credsweeper/deep_scanner/deep_scanner.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (csh519) [reply credsweeper/deep_scanner/deep_scanner.py]: Reverted ee9b6ec.

**Review by babenek — APPROVED** [review]
OK.
["src", "doc"] - only two types. If it will be extended - then Enum should be used.

File: credsweeper/scanner/scanner.py [hunk 7]
```diff
@@ -28,16 +28,16 @@ class Scanner:
 
     TargetGroup = List[Tuple[AnalysisTarget, str, int]]
 
-    def __init__(self, config: Config, rule_path: Optional[str]) -> None:
+    def __init__(self, config: Config, rule_path: Optional[str], usage_list: List[str] = ["src", "doc"]) -> None:
```
Comment (babenek) [inline credsweeper/scanner/scanner.py:31]:
```suggestion
    def __init__(self, config: Config, rule_path: Optional[str], usage_list: Optional[List[str]] = None) -> None:
```

File: credsweeper/scanner/scanner.py [hunk 8]
```diff
@@ -28,16 +28,16 @@ class Scanner:
...
         self.config = config
         self.__scanner_for_rule: Dict[str, Type[ScanType]] = {}
         self.rules: List[Rule] = []
         # init with MAX_LINE_LENGTH before _set_rules
         self.min_pattern_len = MAX_LINE_LENGTH
-        self._set_rules(rule_path)
+        self._set_rules(rule_path, usage_list)
```
Comment (babenek) [inline credsweeper/scanner/scanner.py:37]:
```suggestion
        self._set_rules(rule_path, usage_list if isinstance(usage_list, list) else ["src", "doc"])
```

**Review by babenek — COMMENTED** [review]
Suggested changes to supress pylint warning.

File: credsweeper/app.py [hunk 9]
```diff
@@ -69,38 +70,64 @@ def __init__(self,
...
+    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+    def _get_config_path(self, config_path: Optional[str], config_file: str) -> str:
         if config_path:
-            config_dict = Util.json_load(config_path)
+            return config_path
         else:
             dir_path = os.path.dirname(os.path.realpath(__file__))
-            config_dict = Util.json_load(os.path.join(dir_path, "secret", "config.json"))
+            return os.path.join(dir_path, "secret", config_file)
```
Comment (babenek) [inline credsweeper/app.py:101]: BTW, https://github.com/Samsung/CredSweeper/blob/bc1de2cca32972060d5f227fc4f623e8789517f2/credsweeper/__init__.py#L26

**Review by babenek — DISMISSED** [review]
Good. The warning has vanished.

File: credsweeper/app.py [hunk 10]
```diff
@@ -247,10 +274,12 @@ def file_scan(self, content_provider: Union[DiffContentProvider, TextContentProv
             candidates.append(dummy_candidate)
 
         else:
-            # deep scan with possibly data representation
             if self.config.depth:
-                new_size_limit = self.config.size_limit if self.config.size_limit else RECURSIVE_SCAN_LIMITATION
-                candidates = self.deep_scanner.scan(content_provider, self.config.depth, new_size_limit)
+                # deep scan with possibly data representation
```
Comment (kmnls) [inline credsweeper/app.py:278]:
```suggestion
                # deep scan with possible data representation
```

**Review by xDizzix — DISMISSED** [review]
The document scanning feature looks good.

**Review by xDizzix — APPROVED** [review]
Upd

**Review by babenek — APPROVED** [review]

**Review by kmnls — CHANGES_REQUESTED** [review]
It seems to me that this option completely duplicates the --depth option.
There is not enough description in the code and PR message both about what is the usage difference and why the new option is necessary.

**Comment by csh519** [comment]
> It seems to me that this option completely duplicates the --depth option. There is not enough description in the code and PR message both about what is the usage difference and why the new option is necessary.

@kmnls 

For now CredSweeper just targeted on source code scanning, however internally we got requirements to make CredSweeper to work on documents.
As you know, CredSweeper supports document scanning feature by `--depth` option.
But there is some options which doesn't need `depth` value.
So I want to separate document scanning feature from `depth` option and improve it more, that's why I make this PR for adding `--doc` option.
It's the first step of adding and improving document scanning feature.

And there is one more reason in here, it's about internal issue.
If you want to know about this, please message to me by messenger.

**Review by kmnls — APPROVED** [review]
Agree but further improvements are required.
--doc option is interchanged with the behavior of --depth option. Probably --depth may be made a sub-option of --doc?
As I see, the lookup patterns will have an additional option which controls the mode of use but I could not find a proper test for this feature ("pattern fired/not fired" depending on the scanning mode). Please provide

**Comment by csh519** [comment]
> Agree but further improvements are required. --doc option is interchanged with the behavior of --depth option. Probably --depth may be made a sub-option of --doc? As I see, the lookup patterns will have an additional option which controls the mode of use but I could not find a proper test for this feature ("pattern fired/not fired" depending on the scanning mode). Please provide

@kmnls 

Okay. I'll add reasonable use-cases and tests for `--doc` feature later.
Thank you for your comment.

**Pull request merged**